### PR TITLE
[MRG] Support "all-categorical" space without one-hot-encoding in gp_minimize

### PR DESCRIPTION
**Pull request by MechCoder**
Fixes https://github.com/scikit-optimize/scikit-optimize/issues/56

### CODE REVIEW

**Comment by MechCoder** [comment]
@betatim @glouppe This is ready for reviews. Categorical input is now supported, basically

        X = np.asarray([
        ["ham", "spam", "ted"],
        ["ham", "ted", "ted"],
        ["ham", "spam", "spam"]])
       y = rng.randn(3)
       hm = HammingKernel(length_scale=[1.0, 1.0, 1.0])

       gpr = GaussianProcessRegressor(hm)
       gpr.fit(X, y)
       gpr.predict(X)

**Comment by glouppe** [comment]
General comment, can you put the kernel code under `skopt.learning`?

**Comment by glouppe** [comment]
Documentation comment, could you also switch to the docstring format I used everywhere else? (This one does not render properly in Markdown)

File: skopt/kernels/hamming_kernel.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+from sklearn.gaussian_process.kernels import Kernel
+from sklearn.gaussian_process.kernels import NormalizedKernelMixin
+from sklearn.gaussian_process.kernels import StationaryKernelMixin
+
+class HammingKernel(StationaryKernelMixin, NormalizedKernelMixin, Kernel):
+    """
+    The HammingKernel is used to handle categorical inputs.
+
+    ``K(x_1, x_2)`` is given by \sum_{j=1}^{d}exp(-ls_j * (I(x_1j != x_2j)))
+    If all the categorical features of two samples are same then
```
Comment (glouppe) [inline skopt/kernels/hamming_kernel.py]: are _the_ same

File: skopt/kernels/hamming_kernel.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
+from math import exp
```
Comment (glouppe) [inline skopt/kernels/hamming_kernel.py]: is there a reason for using `math.exp` instead of `np.exp`?

**Comment by glouppe** [comment]
Thanks! This is a nice first step towards support of categorical variables. This assumes that all features are categorial though, doesn't it? (i.e., we cannot mix categorical and continuous features)

**Comment by MechCoder** [comment]
Actually I ran into a bug while testing the kernel gradient with respect to hyperparameters. I have fixed it in this PR and I have reported it here (https://github.com/scikit-learn/scikit-learn/issues/6701)

**Comment by MechCoder** [comment]
> This assumes that all features are categorial though, doesn't it?

Yes, but we can address the mixed features type in a different PR
In 4.1.2 of http://arxiv.org/pdf/1211.0906v2.pdf, they just advise to find `K` for categorical and continuous features separately and do a pointwise multiplication.

Also, could you make another pass?

**Comment by glouppe** [comment]
This should be rebased on top of master. Then I assume this special kernel can be used when all dimensions are `Categorical`, right? (Not sure what do to for mixed types)

**Comment by MechCoder** [comment]
For mixed-types we can have something as described here https://github.com/scikit-learn/scikit-learn/issues/6868 

(I'll try to address that today)

**Comment by MechCoder** [comment]
@glouppe This does not refactor `gp_optimize` but just adds the kernel. I'm not sure why I should rebase. Could you clarify?

**Comment by glouppe** [comment]
Well indeed! I was just assuming you would also make the necessary changes to `gp_optimize`.

**Comment by MechCoder** [comment]
Let us do incremental development :P

**Comment by MechCoder** [comment]
I started working on the `SelectDimensionKernel` here (https://github.com/MechCoder/scikit-learn/tree/select_kernel)

**Comment by MechCoder** [comment]
Just a FYI, we should push this further due to non-trivial issues in sklearn's implementation.

https://github.com/scikit-learn/scikit-learn/issues/6961
https://github.com/scikit-learn/scikit-learn/issues/6868

**Comment by glouppe** [comment]
Do you wish to finish this for 0.1?

**Comment by MechCoder** [comment]
(That moment when I look back and cannot believe I wrote such pathetic code) :-/

**Comment by MechCoder** [comment]
I still have to write the `gradient_x` method and integrate it with the optimizer functions. Will continue on that tomorrow.

**Comment by MechCoder** [comment]
@glouppe @betatim This is ready for reviews!

File: skopt/optimizer/optimizer.py [hunk 3]
```diff
@@ -195,6 +196,7 @@ def tell(self, x, y, fit=True):
             only be fitted after `n_random_starts` points have been queried
             irrespective of the value of `fit`.
         """
+        is_categorical = all([isinstance(dim, Categorical) for dim in self.space.dimensions])
```
Comment (glouppe) [inline skopt/optimizer/optimizer.py]: Shouldnt this be set once at init instead? Or better, as propery of the `space` instance?

File: skopt/space/space.py [hunk 4]
```diff
@@ -46,6 +46,9 @@ def check_dimension(dimension, transform=None):
         the returned Dimension instance the values are scaled
         between 0 and 1.
 
+        If transform is set to "Identity", when `transform` is called
```
Comment (glouppe) [inline skopt/space/space.py]: `"identity"` (without the cap)

Comment (glouppe) [inline skopt/space/space.py]: also the accepted values for `transform` need to be updated in the docstring just above (line 44)

File: skopt/space/space.py [hunk 5]
```diff
@@ -273,10 +276,18 @@ def __init__(self, categories, prior=None):
...
             are equally likely.
+
+        * `transform` [string, "onehot" or "identity"]:
+            When `transform` is set to "identity", the transformed space
+            is the same as the original space.
         """
         self.categories = categories
-        self.transformer = CategoricalEncoder()
-        self.transformer.fit(self.categories)
+        self.transform_ = transform
```
Comment (glouppe) [inline skopt/space/space.py]: Not sure I like setting both `self.transform` and `self.transform_`. This is confusing. Only `self.transform` seems necessary.

**Comment by glouppe** [comment]
Also Travis complains.

File: skopt/space/space.py (hunk 5 above)
Comment (MechCoder) [inline skopt/space/space.py]: I did that because `self.transform` is a method.

File: skopt/optimizer/optimizer.py (hunk 3 above)
Comment (MechCoder) [inline skopt/optimizer/optimizer.py]: I need to check once that all are dimensions are Categorical before instantiating the space, no?

The reason being that if all are Categorical, I need to use a different transform.

Comment (MechCoder) [inline skopt/optimizer/optimizer.py]: I did not realize you were commenting on the `optimizer` diff. I've addressed this.

**Comment by MechCoder** [comment]
updated!

**Comment by MechCoder** [comment]
After doing this, we can add different kernels operating on different dimensions in a separate PR.

**Comment by glouppe** [comment]
Thanks for the changes. Merging.

**Comment by MechCoder** [comment]
Thanks! Will try to implement the `SelectDimensionKernel` either tonight or after I get back in a few days.

**Comment by betatim** [comment]
> Thanks! Will try to implement the SelectDimensionKernel either tonight or after I get back in a few days.

new issue so it doesn't get lost?
